core/src/main/scala/com/metamx/tranquility/finagle/FinagleRegistry.scala

```diff
@@ -31,23 +32,60 @@ import com.twitter.util.Future
 import com.twitter.util.Time
 import java.util.concurrent.atomic.AtomicBoolean
 import org.scala_tools.time.Implicits._
+import scala.collection.Set
 import scala.collection.mutable
 
 /**
-  * Registry of shared Finagle HTTP Curator-discovered services. The services can be returned by closing them. When
-  * the last reference to a service has been returned, the service is closed.
+  * Registry of shared Finagle services.
   */
-class FinagleRegistry(config: FinagleRegistryConfig, disco: Disco) extends Logging
+class FinagleRegistry(
+  config: FinagleRegistryConfig,
+  resolvers: Seq[Resolver]
+) extends Logging
 {
-  private[this] val lock     = new AnyRef
-  private[this] val resolver = new DiscoResolver(disco)
-  private[this] val clients  = mutable.HashMap[String, SharedService[http.Request, http.Response]]()
+  // Backwards compatible constructor.
+  def this(config: FinagleRegistryConfig, disco: Disco) = {
+    this(config, Seq(new DiscoResolver(disco)))
+  }
+
+  private val lock        = new AnyRef
+  private val clients     = mutable.HashMap[String, SharedService[http.Request, http.Response]]()
+  private val resolverMap = mutable.HashMap[String, Resolver]()
 
-  private[this] def mkclient(service: String) = {
+  // Add all resolvers from the constructor.
+  resolvers foreach addResolver
+
+  def addResolver(resolver: Resolver): Unit = {
+    lock.synchronized {
+      if (resolverMap.contains(resolver.scheme)) {
+        throw new IllegalArgumentException(s"Already have resolver for scheme[${resolver.scheme}]")
+      }
+
+      log.info(s"Adding resolver for scheme[${resolver.scheme}].")
+      resolverMap.put(resolver.scheme, resolver)
+    }
+  }
+
+  def schemes: Set[String] = lock.synchronized {
+    resolverMap.keySet
+  }
+
+  private def mkid(scheme: String, name: String): String = s"$scheme!$name"
```

Comment:
is `!` a forbidden character?

for example, yahoo is commonly represented as "Yahoo!" which would cause problems here.